src/sm.c

```diff
@@ -0,0 +1,480 @@
+    {
+        /*Codes_SRS_SM_02_007: [ If sm is NULL then sm_open_begin shall fail and return SM_ERROR. ]*/
+        LogError("invalid argument SM_HANDLE sm=%p", sm);
+        result = SM_ERROR;
+    }
+    else
+    {
+        /*Codes_SRS_SM_02_039: [ If the state is not SM_CREATED then sm_open_begin shall return SM_EXEC_REFUSED. ]*/
+        LONG state = InterlockedAdd(&sm->state, 0);
+        if ((state & SM_STATE_MASK) != SM_CREATED)
+        {
+            LogError("sm name=%s. Cannot sm_open_begin that which is in %" PRI_SM_STATE " state", sm->name, SM_STATE_VALUE(state));
+            result = SM_EXEC_REFUSED;
+        }
+        else
+        {
+            /*Codes_SRS_SM_02_040: [ sm_open_begin shall switch the state to SM_OPENING. ]*/
+            if (InterlockedCompareExchange(&sm->state, state - SM_CREATED + SM_OPENING + SM_STATE_INCREMENT, state) != state)
+            {
+                LogError("sm name=%s. sm_open_begin state changed meanwhile (it was %" PRI_SM_STATE "). likely competing threads.", sm->name, SM_STATE_VALUE(state));
+                result = SM_EXEC_REFUSED;
+            }
+            else
+            {
+                /*Codes_SRS_SM_02_009: [ sm_open_begin shall return SM_EXEC_GRANTED. ]*/
+                result = SM_EXEC_GRANTED;
+            }
+        }
+    }
+    return result;
+}
+
+void sm_open_end(SM_HANDLE sm)
+{
+    /*Codes_SRS_SM_02_010: [ If sm is NULL then sm_open_end shall return. ]*/
+    if (sm == NULL)
+    {
+        LogError("invalid argument SM_HANDLE sm=%p", sm);
+    }
+    else
+    {
+        /*Codes_SRS_SM_02_041: [ If state is not SM_OPENING then sm_open_end shall return. ]*/
+        LONG state = InterlockedAdd(&sm->state, 0);
+        if ((state & SM_STATE_MASK) != SM_OPENING)
+        {
+            LogError("sm name=%s. cannot sm_open_end that which is in %" PRI_SM_STATE " state", sm->name, SM_STATE_VALUE(state));
+        }
+        else
+        {
+            /*Codes_SRS_SM_02_042: [ sm_open_end shall switch the state to SM_OPENED. ]*/
+            if (InterlockedCompareExchange(&sm->state, state - SM_OPENING + SM_OPENED + SM_STATE_INCREMENT, state) != state)
+            {
+                LogError("sm name=%s. sm_open_end state changed meanwhile (it was%" PRI_SM_STATE ", likely competing threads.", sm->name, SM_STATE_VALUE(state));
+            }
+            else
+            {
+                /*we are done*/
+            }
+        }
+    }
+}
+
+static SM_RESULT sm_close_begin_internal(SM_HANDLE sm)
+{
+    SM_RESULT result;
+
+    LONG state;
+    /*Codes_SRS_SM_02_045: [ sm_close_begin shall set SM_CLOSE_BIT to 1. ]*/
+    if (((state=InterlockedOr(&sm->state, SM_CLOSE_BIT)) & SM_CLOSE_BIT) == SM_CLOSE_BIT)
+    {
+        /*Codes_SRS_SM_02_046: [ If SM_CLOSE_BIT was already 1 then sm_close_begin shall return SM_EXEC_REFUSED. ]*/
+        LogError("sm name=%s. another thread is performing close (state=%" PRI_SM_STATE ")", sm->name, SM_STATE_VALUE(state));
```

Comment:
>name=%s. [](start = 21, length = 8)

Love this

Reply:
:)

---
In reply to: [423410580](https://github.com/Azure/azure-c-util/pull/37#discussion_r423410580) [](ancestors = 423410580)